msteams/provider/client.py

```diff
@@ -0,0 +1,253 @@
+import base64
+import asyncio
+import aiohttp
+
+import requests
+from flask import current_app as app, request
+from msal import ConfidentialClientApplication
+
+from . import UpstreamProviderError
+
+AUTHORIZATION_HEADER = "Authorization"
+BEARER_PREFIX = "Bearer "
+CACHE_SIZE = 256
+
+
+class MsTeamsClient:
+    DEFAULT_SCOPES = ["https://graph.microsoft.com/.default"]
+    SEARCH_URL = "https://graph.microsoft.com/v1.0/search/query"
+    APPLICATION_AUTH = "application"
+    DELEGATED_AUTH = "user"
+
+    def __init__(self, auth_type, search_limit=5):
+        self.access_token = None
+        self.user = None
+        self.auth_type = auth_type
+        self.search_limit = search_limit
+        self.session = None
+        self.loop = None
+        self._start_session()
+
+    def set_user(self, user):
+        self.user = user
+
+    def get_auth_type(self):
+        return self.auth_type
+
+    def _close_loop(self):
+        self.loop.stop()
+        self.loop.close()
+
+    def _start_session(self):
+        self.loop = asyncio.new_event_loop()
+        self.session = aiohttp.ClientSession(loop=self.loop)
+
+    async def _close_session(self):
+        await self.session.close()
+
+    def set_app_access_token(self, tenant_id, client_id, client_secret):
+        try:
+            credential = ConfidentialClientApplication(
+                client_id=client_id,
+                client_credential=client_secret,
+                authority=f"https://login.microsoftonline.com/{tenant_id}",
+            )
+
+            token_response = credential.acquire_token_for_client(
+                scopes=self.DEFAULT_SCOPES,
+            )
+            if "access_token" not in token_response:
+                raise UpstreamProviderError(
+                    "Error while retrieving access token from Microsoft Graph API"
+                )
+            self.access_token = token_response["access_token"]
+        except Exception as e:
+            raise UpstreamProviderError(
+                f"Error while initializing Teams client: {str(e)}"
+            )
+
+    def set_user_access_token(self, token):
+        self.access_token = token
+        self.headers = {"Authorization": f"Bearer {self.access_token}"}
+
+    async def _gather_messages(self, hits):
+        messages = [self._get_message(hit) for hit in hits]
+        return await asyncio.gather(*messages)
+
+    async def _get_message(self, hit):
+        url = f"https://graph.microsoft.com/v1.0/chats/{hit['resource']['chatId']}/messages/{hit['resource']['id']}"
+        params = {
+            "$select": "id,subject,summary,body,from,createdDateTime,webUrl,attachments,eventDetail",
+        }
+
+        async with self.session.get(
+            url,
+            headers={"Authorization": f"Bearer {self.access_token}"},
```

Comment:
In `set_user_access_token()` `self.headers` gets set, but in `set_app_access_token()` it doesn't, can we add that to the app auth and then everywhere the headers are needed you can use `self.headers` instead?

Reply:
Good point - thanks. Done